prusti-interface/src/environment/polonius_info.rs

```diff
@@ -1029,25 +1065,56 @@ impl<'a, 'tcx: 'a> PoloniusInfo<'a, 'tcx> {
     pub fn get_loan_places(&self, loan: &facts::Loan) -> Option<LoanPlaces<'tcx>> {
-        let loan_location = self.loan_position[loan];
-        let mir_block = &self.mir[loan_location.block];
-        if loan_location.statement_index < mir_block.statements.len() {
-            let mir_stmt = &mir_block.statements[loan_location.statement_index];
-            match mir_stmt.kind {
-                mir::StatementKind::Assign(ref lhs_place, ref rvalue) => Some(LoanPlaces {
-                    dest: lhs_place.clone(),
-                    source: rvalue.clone(),
-                    location: loan_location,
-                }),
-
-                ref x => {
-                    debug!("Borrow starts at statement {:?}", x);
-                    None
-                }
+        let location = self.loan_position[loan];
+        let assignment = self.get_assignment_for_loan(*loan)?;
+        let (dest, source) = assignment.as_assign().unwrap();
+        let dest = dest.clone();
+        let source = source.clone();
+        Some(LoanPlaces { dest, source, location })
+    }
+
+    /// Returns the atomic assignment that created a loan.
+    pub fn get_assignment_for_loan(&self, loan: facts::Loan) -> Option<mir::Statement<'tcx>> {
+        let location = self.loan_position[&loan];
+        let stmt = self.mir.statement_at(location)?.clone();
+        let mut assignments = stmt.split_assignment(&self.mir);
+
+        // TODO: This is a workaround. It seems like some local variables don't have a local variable declaration in the MIR. One example of this can be observed in the implementation of the clone method in `prusti/tests/verify/pass/copy/generics.rs`. We don't have region information for these local variables, which is why the code below would drop all assignments and return `None`. So if there is only one possible assignment that may have created a loan, we return it directly without inspecting the region information. Of course once a local variable that is a tuple is not declared explicitly, this will fail again.
+        if assignments.len() == 1 {
+            return assignments.pop();
+        }
+
+        let regions = self.get_loan_regions(loan);
+
+        // Drops all assignments that don't share a region with the loan we're interested in.
+        assignments.retain(|stmt| {
+            let (lhs, _) = stmt.as_assign().unwrap_or_else(||
+                unreachable!("Borrow starts at statement {:?}", stmt));
+
+            if let Some(region) = self.place_regions.for_place(lhs) {
+                regions.contains(&region)
+            } else {
+                false
             }
-        } else {
-            None
+        });
+
+        if assignments.len() > 1 {
+            unreachable!(
+                "Cannot determine assignment for loan {:?}, because there are multiple possible options: {:?}",
+                loan, assignments);
         }
+
+        assignments.pop()
+    }
```

Comment:
The idea here is to split the assignment that created a loan, for example

    _3 = (move _4, move _5)

into its atomic parts, which are

    _3.0 = move _4
    _3.1 = move _5

in this case. The LHS of the atomic assignment that created the loan must have a region that requires the terms of the loan to be enforced. If there is exactly one such atomic assignment, we can return it. If there are more (I'm not sure whether this can happen), we panic.

I'm not sure whether this works in general.